### DISCUSSION TITLE
Possible to get Optimal solution only for CPSat/MIP solver?

### DISCUSSION THREAD

**Discussion by aihiangng** [google/or-tools#2945] [CP-SAT questions] [answered]
For a given problem, if I know for sure there is an optimal solution. Is there anyway I can set up the solver such that it gives me the optimal and not feasible solution? The CPSat model currently will give optimal or feasible solution, however I need the optimal one, anyway i can achieve this?

![image](https://user-images.githubusercontent.com/28476498/144540680-57af5d52-28d8-4f1f-9558-bcaf7f1f2543.png)

It says here that CPSat will focus on the constraints rather than objective function. Does this mean it will return a feasible solution the moment one is found and may not return me the optimal solution even if there is one? If yes, can I set it such that it will always return only the optimal solution ?

**Comment by lperron** [comment, marked as answer]
1) You are aware that these are NP-Hard problem, that can take millions of years to prove optimality ? So the expectation that you can 'demand' optimality is just absurd.

2) The part on wikipedia does not apply to CP-SAT. The text is already opinionated and mostly biased in favor of MIP solvers.  In that regard, CP-SAT is a (Non mixed) MIP solver and a good one at that.

**Reply by aihiangng** [reply]
From the Google OR CP Sat solver shown here, https://developers.google.com/optimization/cp/cp_solver, we can do a check on the solution

![image](https://user-images.githubusercontent.com/28476498/144770725-2d4f98be-264c-408d-ad8d-2105ef2f2f84.png)


The solution can be optimal (feasible), feasible, infeasible etc... If I am working with an assignment problem with CP Sat with an objective function, and I know there is an optimal solution*, does that mean CP Sat will always give optimal? Or does it not guarantee and optimal solution. Meaning the once the solver managed to find a feasible solution it will not  not bother to continue searching for an optimal solution? 

*I know that there is an optimal solution by using the example of a very simplified assignment problem that I can check manually. 

The gist of my qn is, if I know theres an optimal solution, will the solver give me an optimal solution OR will it give me any feasible solution it reaches first.

**Reply by aihiangng** [reply]
From the google OR notebook here https://github.com/google/or-tools/blob/stable/ortools/sat/doc/README.md#searching-for-one-optimal-solution-of-a-cp-sat-model

It states that "By default, searching for one solution will return the first solution found if the model has no objective, or the optimal solution if the model has an objective."

Does this mean with an objective function, CPSat will always return an optimal solution?

**Reply by lperron** [reply]
have you read my answer ? It will if it can.
